### ISSUE TITLE
Problem in stage-out of large files for multi-node execution

### ISSUE THREAD

**Issue by clmendes** [LLNL/UnifyFS#540]
As I had reported in the call yesterday, I'm facing a problem when exporting a large file from /unifyfs back to persistent storage, after a program finishes, if the execution was done on multiple nodes.

I reproduced the problem with the program below, executing it with four processors on four nodes of Quartz at Livermore. The produced file, which should have 64 MBytes, ends up much smaller when four nodes are used for the execution. When I execute it with four processors on the same node, everything works fine, and the exported file does have 64 MB. 

In this program, all ranks write to the same file, at different positions according to their rank. The total file size, in MB, is always the same, as dictated by symbol SIZEMB.

In my job script, I have the following line near the end:

$UNIFYFS_INSTALL/bin/unifyfs terminate -o outfile -S $mydir

and in that outfile I have

/unifyfs/datafile   /g/g12/mendes3/ ... /datafile

Like I said above, running within one node, everything works fine. Also, if I reduce the filesize to 4 MB (by changing SIZEMB to 4 in the source code), it also works, even with four nodes!

In addition to result in a shorter file when running on four nodes, another problem is that even in that case, I end up with a file called unifyfs-stage.status in my dir that contains "success"! Thus, it reports success, but the file is shorter than expected!

```
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>
#include <unifyfs.h>
#define SIZEMB 64    /* Total file size in MB */

int main(int argc, char **argv)
{
    int i, nranks, ret, nbytes, myrank;
    char *buf;
    MPI_File fh;
    MPI_Status status;
    MPI_Offset offset;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
    MPI_Comm_size(MPI_COMM_WORLD, &nranks);

    /* Mount UnifyFS */
    ret = unifyfs_mount("/unifyfs", myrank, nranks, 0);
    if (ret) {
        fprintf(stderr,"[%d] unifyfs_mount failed (return = %d)\n", myrank, ret);
        exit(-1);
    }

    /* Initialize data */
    nbytes = (SIZEMB * 1024 * 1024) / nranks;
    buf = malloc(nbytes);
    for (i=0; i<nbytes; i++) buf[i] = (char) myrank;

    /* Open File */
    ret = MPI_File_open(MPI_COMM_WORLD, "ufs:/unifyfs/datafile",
                            MPI_MODE_CREATE | MPI_MODE_RDWR, MPI_INFO_NULL, &fh);
    if (ret != MPI_SUCCESS) fprintf(stderr,"[%d] Open error: %d\n", myrank, ret);

    /* Write data to file */
    offset = myrank * nbytes;
    ret = MPI_File_write_at(fh, offset, buf, nbytes, MPI_BYTE, &status);
    if (ret != MPI_SUCCESS) fprintf(stderr,"[%d] Write error: %d\n", myrank, ret);
    else fprintf(stderr,"[%d] Write Success: offset=%d nbytes=%d\n", myrank, offset, nbytes);

    /* Close file */
    ret = MPI_File_close(&fh);
    if (ret != MPI_SUCCESS) fprintf(stderr,"[%d] Close error: %d\n",myrank,ret);

    MPI_Finalize();
    return 0;
}

```

**Comment by MichaelBrim** [comment]
The test program is missing a call to unifyfs_unmount(). I don't think that's the root cause of the problem, just want to eliminate that possibility.

**Comment by clmendes** [comment]
Good catch @MichaelBrim , I added now a call to unifyfs_unmount() just before the call to MPI_Finalize. And now I do see lines like these in the client log:

2020-07-22T07:57:30 tid=35038 @ unifyfs_unmount() [unifyfs.c:2616] syncing data
2020-07-22T07:57:30 tid=35038 @ unifyfs_unmount() [unifyfs.c:2628] calling unmount

However, the problem persists. When running it on Quartz with 4 processors on 4 nodes: I did two executions, the first one produced a 35 MB file, the second produced a 43 MB file (instead of the expected 64 MB file). In both executions, file unifyfs-stage.status had the contents "success"!

**Comment by sandrain** [comment]
@clmendes I've tested this, and it looks like the `unifyfs` utility fails to correctly coordinate with the `unifyfs-stage` utility, at least in my environment. Just to make sure, could you try to transfer data using the `unifyfs-stage` utility directly (instead of using the `unifyfs terminate` utility. So, instead of:

```
$UNIFYFS_INSTALL/bin/unifyfs terminate -o outfile -S $mydir
```

, try:

```
mpirun/srun [options] $UNIFYFS_INSTALL/libexec/unifyfs-stage outfile
$UNIFYFS_INSTALL/bin/unifyfs terminate
```

And, see if the file is transferred correctly.

In my case, running `unifyfs-stage` directly works. I think we would need to fix the `unifyfs` utility to correctly detect the transfer status. Maybe, we would have to just wait (using `wait(2)` syscall), instead of using the timeout and publishing the status file.

**Comment by clmendes** [comment]
Thanks @sandrain and @craigsteffen , yes, by executing unifyfs-stage separately, i.e. before issuing the terminate command to unifyfs, it works fine! I even increased the filesize to 128MB, used the parallel "-p" option for unifyfs-stage, and ran the job three times in a row, all three executions produced a valid output file (size=128 MB, correct contents).

I'd say that this solves the problem for me. I will avoid using the stage-out feature together with 'terminate' in the future.

**Comment by craigsteffen** [comment]
Thanks Celso.  I'll see if I can reproduce the problem with calling stage
with "unifyfs terminate" on catalyst or anywhere else.  I'll look at where
uniffys calls unifyfs-stage and see if there's something obviously wrong.

We'll reconvene Monday.

Craig

**Comment by sandrain** [comment]
@clmendes Could you check if #541 fixes this? In my test, passing the stage-out option to the `unifyfs` utility works with this fix.

**Comment by clmendes** [comment]
@sandrain and @craigsteffen , yes, I just confirmed that the fix provided by #541 does solve the problem. I can now go back to using the stage-out option with terminate. I tested that with both my example code (posted above), which creates a 128MB file (using SIZEMB=128), and with my original ROMIO test code (coll_perf.c), which creates a 220MB file. Thanks!

**Comment by adammoody** [comment]
Fixed with https://github.com/LLNL/UnifyFS/pull/541
